src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java

```diff
@@ -100,6 +162,14 @@ public boolean processRow(StepMetaInterface smi, StepDataInterface sdi) throws K
 
                 setOutputDone();
 
+                //now output rowset closed, so we can wait for our confirmation steps
+                if ( data.activeConfirmation )  {
+                    while (!isStopped() && data.watchedConfirmStep.size() > 0) {    
+                        Thread.sleep(500);

```

Comment:
yes, it can take much more than 500 ms, it is not wait 500ms for end, it looping here with delay of 500ms.

transFinished is used to close transaction, adn if customer activate Transactional flag, all ack/nack will be performed at the end of transformation. 

current loop is set after OutputDone, so all subsequent steps is not blocked on stream from this CCONSUMER. this loop is just to let ACK/REJECT step to fully process data and notify consumer accrodingly.